application/src/main/java/org/opentripplanner/model/impl/SubmodeMappingModule.java

```diff
@@ -0,0 +1,103 @@
+import org.opentripplanner.routing.graph.Graph;
+import org.opentripplanner.transit.model.basic.TransitMode;
+
+@Module
+public class SubmodeMappingModule implements GraphBuilderModule {
+
+  private static final String INPUT_FEED_TYPE = "Input feed type";
+  private static final String INPUT_LABEL = "Input label";
+  private static final String GTFS_ROUTE_TYPE = "GTFS route type";
+  private static final String NETEX_SUBMODE = "NeTEx submode";
+  private static final String REPLACEMENT_MODE = "Replacement mode";
+  private static final String[] ALL = {
+    INPUT_FEED_TYPE,
+    INPUT_LABEL,
+    GTFS_ROUTE_TYPE,
+    NETEX_SUBMODE,
+    REPLACEMENT_MODE,
+  };
+
+  private final GraphBuilderDataSources graphBuilderDataSources;
+  private final Graph graph;
+
+  public SubmodeMappingModule(GraphBuilderDataSources graphBuilderDataSources, Graph graph) {
+    this.graphBuilderDataSources = graphBuilderDataSources;
+    this.graph = graph;
+  }
+
+  @Provides
+  @Singleton
+  public static SubmodeMappingService submodeMappingService(Graph graph) {
+    return new SubmodeMappingService(graph.submodeMapping);
+  }
+
+  public void read(DataSource dataSource, Map<SubmodeMappingMatcher, SubmodeMappingRow> map) {
+    try {
+      var reader = new CsvReader(dataSource.asInputStream(), StandardCharsets.UTF_8);
+      reader.readHeaders();
+      var headers = reader.getHeaders();
+      for (var header : ALL) {
+        if (Arrays.stream(headers).noneMatch(h -> h.equals(header))) {
+          throw new OtpAppException("submode mapping header not found: " + header);
+        }
+      }
+      while (reader.readRecord()) {
+        var inputFeedType = reader.get(INPUT_FEED_TYPE);
+        var inputLabel = reader.get(INPUT_LABEL);
+        var gtfsRouteType = reader.get(GTFS_ROUTE_TYPE);
+        var netexSubmode = reader.get(NETEX_SUBMODE);
+        var replacementMode = reader.get(REPLACEMENT_MODE);
+        var matcher = new SubmodeMappingMatcher(inputFeedType, inputLabel);
+        var row = new SubmodeMappingRow(
+          Integer.parseInt(gtfsRouteType),
+          netexSubmode,
+          TransitMode.valueOf(replacementMode)
+        );
+        map.put(matcher, row);
+      }
+    } catch (IOException ioe) {
+      throw new OtpAppException("cannot read submode mapping config file", ioe);
+    }
+  }
+
+  public void useDefaultMapping(Map<SubmodeMappingMatcher, SubmodeMappingRow> map) {
+    map.put(
+      new SubmodeMappingMatcher("GTFS", "714"),
+      new SubmodeMappingRow(714, "railReplacementBus", TransitMode.RAIL)
+    );
+    map.put(
+      new SubmodeMappingMatcher("NeTEx", "railreplacementBus"),
+      new SubmodeMappingRow(714, "railReplacementBus", TransitMode.RAIL)
+    );
+  }
+
+  @Override
+  public void buildGraph() {
+    var dataSources = graphBuilderDataSources.getSubmodeMappingDataSource();
+    var map = new HashMap<SubmodeMappingMatcher, SubmodeMappingRow>();
+    boolean mapWritten = false;
```

Comment:
Could this boolean be replaced with just if (!dataSources.isEmpty()) ... else ..., or by a ternary operation.